### ISSUE TITLE
Spanish ISO keyboard layout: BUG / Keyboard shortcuts could be considered 'mislabeled' on foreign keyboard layouts

### ISSUE THREAD

**Issue by Jivanvl** [Microsoft/vscode#24692]
# Regression / Bug

- VSCode Version:
Version 1.11.1 (1.11.1)
- OS Version:
macOS Sierra v 10.12.4

Steps to Reproduce:

1. Connect a foreign keyboard layout (in my case a Spanish ISO keyboard layout)
2. Try to toggle the line comments using `shift + cmd + &` and it won't work, but using `shift + cmd + /` works. 

I think there might be some confusion as the Spanish keyboard layouts have the `&` on the `6` key while US Keyboard style layouts have the `&` on the `7` key

**Comment by crazyfraggle** [comment]
I'm seeing similar issue on Linux:

Native keyboard layout is Norwegian, chosen keyboard layout is en-US.

When trying to toggle line comments by pressing Ctrl-/ (en-US layout, actually Ctrl-- on physical keyboard) it zooms out instead. This goes for every keyboard shortcut where the character is differently located from the physical keyboard. Including Ctrl-`, Ctrl-+, etc.

While typing in editor, chosen layout is respected, which adds to muscle memory confusion.

**Comment by danielrob** [comment]
I can confirm this is a bug + regression, and provide a workaround for my case. 

I use a french canadian keyboard layout in OSX El Capitain with a native keyboard english layout. Previously it did not matter which language I was in, I could press the native keys `cmd` + `/` to toggle a comment, even though in French Canadian layout this corresponds to `cmd` + `é`. 

The workaround in my case is: while in french canadian keyboard - re-set the mapping  for toggle comments in keyboard shortcuts to the native keys to `cmd` + `/` (i.e. what it always was). Interestingly when doing this, it will update the display to `cmd` + `é`, and when switching to an english keyboard this will be updated to `cmd` + `/`, and vice versa. Whereas, originally, regardless of what language I was in `cmd` + `/` was displayed.

**Comment by marcoqu** [comment]
Same for Italian layout: toggle comments was `ctrl+ò`, toggle terminal was `ctrl+ù`.
Now both don't work and require editing of keyboard shortcuts.
Btw, I'm on windows 10, not on mac.

**Comment by alexandrudima** [comment]
This bug will continue tracking the OP's problem (Spanish ISO keyboard layout + mac) which I will investigate when I get my hands on a mac (soonish).

---

@crazyfraggle @danielrob 

It appears you want something different. You want a keyboard shortcut that remains constant irrespective of your keyboard layout. Luckily, on Mac or Linux you can do this if you write your bindings using scan codes. e.g.:
```
{ "key": "cmd+[Slash]", "command": "command": "editor.action.commentLine", "when": "editorTextFocus"
```

In 1.11 we've added that we detect when the keyboard layout is changed while VS Code is running. The reason this worked before as you would desire it was a bug, where we would detect the keyboard layout only on startup.

You can read more about our current keybinding dispatching at https://github.com/Microsoft/vscode/wiki/Keybindings

---

@marcoqu Keyboard handling is implemented very differently on Windows or on Mac. An explanation is available at https://github.com/Microsoft/vscode/wiki/Keybindings .  Please create a separate issue and I will investigate.

---

**Comment by danielrob** [comment]
@alexandrudima I can raise another bug perhaps, but what you stated is contradictory. 
The 1.11 changes are good: now, when I switch keyboards, the display key is updated according to the new mapping (but the physical key remains the same). 

However this doesn't actually work for a default install of VS Code. Regardless of startup language, the display key is always the english keyboard equivalent, even while changing languages while VS Code is running. It is only after I manually re-set that key that VS Code behaves as you described. And this also breaks the default key mappings if I use a non-english keyboard. (i.e first time install, open up with french keyboard layout, and pressing `cmd` + `0xBF` does not toggle comments anymore like it used to).

**Comment by alexandrudima** [comment]
@danielrob

It is expected that there are differences between VS Code 1.10 and 1.11. That is because in 1.10 on OSX keybinding dispatching was done on `keyCode` and on 1.11 keybinding dispatching is done on `code`.

Our change in 1.11 does not mean that the new default keybindings are necessarily better and I am not arguing that at all. Let me try to explain why.

Our source code defines default keybindings in terms of key codes. More specifically, they are optimized for the EN-US keyboard layout. **Nowhere in our source code do we define keyboard layout specific default keybindings**. To pick some examples:
* select all is bound to `cmd+A`
* undo is bound to `cmd+Z`
* comment line is bound to `cmd+/`
* toggle terminal is bound to `` cmd+` ``
* indent lines is bound to `` cmd+] ``

I don't have access to a mac right now, so I'm going with the keyboard layout shown at https://support.apple.com/en-us/HT201794 under French - (F).

If we are to interpret the default bindings as scan codes, then on a French keyboard the default shortcuts would be:
* select all -> `cmd+[KeyA]` -> `cmd+Q` **bad**
* undo -> `cmd+[KeyZ]` -> `cmd+W` **bad**
* comment line -> `cmd+[Slash]` -> `cmd+=` **bad** (people would expect zoom in to be on the key with +)
* toggle terminal -> `cmd+[Backquote]` -> `cmd +@` **good**
* indent lines -> `cmd+[BracketRight]` -> `cmd+$` **neutral**

The only way to get good default keybindings for a specific keyboard layout is if we manually optimize for it. We cannot generate very good defaults for any keyboard layout out of the EN-US layout, this is something you can upvote in #1240.

Your usage pattern appears to be that you constantly switch between keyboard layouts very often. Please keep in mind that there is no proof this is a general usecase. Most folks I talk to (from other issues) have a keyboard layout and appear to simply stick with it. And they want VS Code to work great with that keyboard layout.

**Comment by danielrob** [comment]
@alexandrudima thanks for this explanation, it was really clear, and does explain the differences between 1.10 & 1.11. The only incorrect part was my use case - while I mentioned switching keyboards to illustrate the problem - I only ever program in French Canadian (C) from your link. I think based off your explanation that basically with `keyCode` I had `cmd  + [Slash]` -> `cmd + /`, but now using `code` it is `cmd + [Slash]` -> `cmd + é`. Therefore it required me to reset the keymapping manually, whereas previously this was never required. It appears #1240 would indeed solve this problem.

**Comment by alexandrudima** [comment]
@Jivanvl 

I have just verified on a mac mini using both an ANSI and an ISO keyboard attached to it.

The Toggle Line Comment action is bount to `cmd + shift + 7` by default. This makes sense to me as the keyboard viewer confirms that shift + 7 produces a `/` on the Spanish ISO keyboard layout. I therefore think we do the correct thing.

**Comment by Jivanvl** [comment]
@alexandrudima Thank you for double checking I think I'm going to take another look at my config file, I might've changed something
